Issue: Tester Node — Environment Setup Failures Route to Planner for Auto-Fix

### Problem

When the test environment is broken (missing dependencies, `pytest` not found,
`ModuleNotFoundError`, etc.), `tester_node` currently hangs or enters an infinite
retry loop. The workflow never advances.

The root cause is a **missing responsibility boundary**: the tester tries to handle
environment failures itself — but that's not its job. The tester's job is:
- Write tests
- Run tests
- Analyse results

Environment setup is the **planner's** job.

### Correct Responsibility Split

| Agent | Responsibility |
|---|---|
| **Tester** | Write, run, and analyse tests. Classify failure type. Report environment failures upward — never try to fix them. |
| **Planner** | Receive the env failure report, diagnose the missing dependency, create a fix item (install package, update `pyproject.toml`, create `package.json`, etc.), and re-queue the tester. |
| **Coder** | Execute the planner's fix item (run `pip install`, update config files, etc.) |

### Failure Classification in `tester_node`

The tester must distinguish between two failure types and return a structured report:

```
RESULT: ENV_FAILURE
REASON: No module named 'fastapi'
COMMAND: pytest tests/ -v
OUTPUT: <raw output>
```

vs.

```
RESULT: TEST_FAILURE
FAILED: tests/test_api.py::test_health — AssertionError
OUTPUT: <raw output>
```

**Environment failure patterns** (tester detects, does not fix):
- `command not found` / `'pytest' is not recognized`
- `No module named X` / `ModuleNotFoundError` / `ImportError`
- `Cannot find module 'X'` (Node.js)
- `error: externally-managed-environment`
- `The term 'pytest' is not recognized` (PowerShell)
- Timeout on test command

### New Workflow Route: `tester → planner_env_fix → coder → tester`

```
tester detects ENV_FAILURE
    ↓
planner_env_fix_node (new lightweight node)
    - Reads the failure report
    - Creates a single TodoItem: "Install missing dependency / fix env"
    - Assigns to coder (pip install / npm install / update pyproject.toml)
    ↓
coder executes fix
    ↓
tester retries (same original item)
```

**Safety cap:** maximum 2 env-fix rounds per item. After 2 failed fix attempts,
route to human gate with `stop_reason = "env_setup_failed"` and a clear message
explaining what was tried.

### Implementation

**1. `_classify_test_output(output: str)` in `nodes.py`**

```python
ENV_FAILURE_PATTERNS = [...]  # regexes as listed above

def _classify_test_output(output: str) -> Literal["pass", "test_failure", "env_failure"]:
    if _is_env_failure(output):
        return "env_failure"
    if _is_test_pass(output):
        return "pass"
    return "test_failure"
```

**2. `tester_node` returns structured result**

On env failure, set `phase = WorkflowPhase.ENV_FIXING` and emit a status message
explaining the problem. Does NOT route back to coder directly.

**3. New `planner_env_fix_node` in `nodes.py`**

Lightweight node — no full replanning. Reads `last_test_output`, identifies the
missing package, prepends a single fix TodoItem to the front of the current plan,
routes to coder.

**4. New `WorkflowPhase.ENV_FIXING` in `state.py`**

**5. New `env_fix_attempts: int = 0` field in `GraphState`**

Incremented on each env fix round. When `>= 2`, route to human gate.

**6. Orchestrator wiring**

```
tester → (pass → decide, test_failure → coder, env_failure → planner_env_fix)
planner_env_fix → coder
coder → tester  (when phase == ENV_FIXING, skip peer_review and go straight back)
```

**7. Visible status messages in UI**

- On env failure: `⚠️ Tester: missing dependency detected — handing to planner for auto-fix`
- On fix success retry: `🔄 Retrying tests after environment fix`
- On cap exceeded: `❌ Environment setup failed after 2 attempts — human intervention required`

### Acceptance Criteria

- [ ] `pytest` not found → planner creates `pip install pytest` item → coder runs it → tester retries
- [ ] `ModuleNotFoundError: No module named 'X'` → planner creates install item for `X`
- [ ] `Cannot find module 'X'` (Node) → planner creates `npm install X` item
- [ ] After 2 failed env-fix attempts, route to human gate with `stop_reason = "env_setup_failed"`
- [ ] Actual test assertion failures still route back to coder unchanged
- [ ] Tests passing still advances to human gate unchanged
- [ ] Env-fix round does NOT go through peer review (it's a mechanical fix, not a coding task)
- [ ] All existing tests pass
- [ ] New tests: `tests/test_tester_env_failure.py` with ≥ 12 tests

### Affected Files

- `app/core/nodes.py` — `tester_node`, `_classify_test_output()`, new `planner_env_fix_node`
- `app/core/orchestrator.py` — new routes for `env_failure`, `ENV_FIXING` phase
- `app/core/state.py` — `WorkflowPhase.ENV_FIXING`, `env_fix_attempts: int = 0`
- `tests/test_tester_env_failure.py` — new test file

### Dependencies

None.

--- a/app/core/nodes.py
+++ b/app/core/nodes.py
@@ -2156,8 +2156,68 @@ def planner_review_node(state: GraphState) -> dict:
 # NODE: tester
 # ---------------------------------------------------------------------------
 
+# Patterns in test/runner output that indicate an environment problem,
+# not a test failure. The tester detects these and hands off to planner_env_fix.
+_ENV_FAILURE_PATTERNS: list[re.Pattern[str]] = [
+    re.compile(r"command not found", re.IGNORECASE),
+    re.compile(r"'pytest'\s+is not recognized", re.IGNORECASE),
+    re.compile(r'The term ["\']pytest["\'] is not recognized', re.IGNORECASE),
+    re.compile(r"No module named\s+\S+", re.IGNORECASE),
+    re.compile(r"ModuleNotFoundError", re.IGNORECASE),
+    re.compile(r"ImportError", re.IGNORECASE),
+    re.compile(r"cannot import name", re.IGNORECASE),
+    re.compile(r"Cannot find module", re.IGNORECASE),          # Node.js
+    re.compile(r"error: externally-managed-environment", re.IGNORECASE),
+    re.compile(r"Could not find.*executable", re.IGNORECASE),
+    re.compile(r"No such file or directory.*python", re.IGNORECASE),
+    re.compile(r"python.*not found", re.IGNORECASE),
+    re.compile(r"node.*not found", re.IGNORECASE),
+]
+
+_TEST_PASS_PATTERNS: list[re.Pattern[str]] = [
+    re.compile(r"\d+ passed", re.IGNORECASE),
+    re.compile(r"All tests passed", re.IGNORECASE),
+    re.compile(r"Tests passed", re.IGNORECASE),
+    re.compile(r"OK\s*$", re.MULTILINE),
+]
+
+# Maximum env-fix rounds before giving up and routing to human gate
+_MAX_ENV_FIX_ATTEMPTS = 2
+
+
+def _is_env_failure(output: str) -> bool:
+    """Return True if the output indicates a missing/broken test environment."""
+    return any(pat.search(output) for pat in _ENV_FAILURE_PATTERNS)
+
+
+def _is_test_pass(output: str) -> bool:
+    """Return True if the output clearly indicates all tests passed (no failures)."""
+    has_failure = re.search(r"\d+\s+failed", output, re.IGNORECASE)
+    if has_failure:
+        return False
+    return any(pat.search(output) for pat in _TEST_PASS_PATTERNS)
+
+
+def _classify_test_output(output: str) -> str:
+    """Classify raw test runner output as 'pass', 'env_failure', or 'test_failure'."""
+    if _is_env_failure(output):
+        return "env_failure"
+    if _is_test_pass(output):
+        return "pass"
+    return "test_failure"
+
+
 def tester_node(state: GraphState) -> dict:
-    """Run tests and verification."""
+    """Run tests and verification.
+
+    Classifies the test runner output into three categories:
+    - pass        → advance to decide/human-gate
+    - test_failure → route back to coder for a fix
+    - env_failure  → route to planner_env_fix (never handled here)
+
+    The tester's job is to write, run, and analyse tests.
+    It never modifies the environment — that is the planner's responsibility.
+    """
     item = state.current_item
     if not item:
         emit_error("system", "No item to test")
@@ -2183,22 +2243,59 @@ def tester_node(state: GraphState) -> dict:
         prompt += f"\n\n{intelligence_ctx}"
     prompt += (
         "\nRun all tests, linters, and verification commands. "
-        "Produce a structured test report with PASS or FAIL verdict."
+        "Produce a structured test report with PASS or FAIL verdict. "
+        "If the test runner itself cannot start (missing interpreter, missing module, "
+        "command not found), report that clearly — do NOT attempt to install packages."
     )
 
     result = _invoke_agent("tester", [HumanMessage(content=prompt)], TESTER_TOOLS)
 
-    verdict = "PASS" if "PASS" in result.upper() and "FAIL" not in result.upper() else "FAIL"
+    # -- Classify the raw LLM output ----------------------------------------
     if "**Verdict**: PASS" in result or "Verdict: PASS" in result:
-        verdict = "PASS"
+        llm_verdict = "PASS"
     elif "**Verdict**: FAIL" in result or "Verdict: FAIL" in result:
-        verdict = "FAIL"
+        llm_verdict = "FAIL"
+    elif "PASS" in result.upper() and "FAIL" not in result.upper():
+        llm_verdict = "PASS"
+    else:
+        llm_verdict = "FAIL"
+
+    # Override: if output signals env failure, trust that over LLM verdict
+    classification = _classify_test_output(result)
 
     item.test_report = result
 
-    emit_verdict("tester", verdict, detail=result, item_id=item.id)
+    # -- ENV FAILURE: hand off to planner_env_fix ---------------------------
+    if classification == "env_failure":
+        if state.env_fix_attempts >= _MAX_ENV_FIX_ATTEMPTS:
+            msg = (
+                f"Environment setup failed after {state.env_fix_attempts} fix attempt(s). "
+                "Human intervention required."
+            )
+            emit_status("tester", f"❌ {msg}", **_progress_meta(state, "stopped"))
+            emit_node_end("tester", "Testing", msg)
+            return {
+                "last_test_result": result,
+                "stop_reason": "env_setup_failed",
+                "phase": WorkflowPhase.STOPPED,
+            }
 
-    if verdict == "FAIL":
+        emit_status(
+            "tester",
+            "⚠️ Tester: missing dependency or broken environment detected "
+            "— handing to planner for auto-fix",
+            **_progress_meta(state, "env_fixing"),
+        )
+        emit_node_end("tester", "Testing", "ENV_FAILURE — routing to planner_env_fix")
+        return {
+            "last_test_result": result,
+            "phase": WorkflowPhase.ENV_FIXING,
+        }
+
+    # -- Normal PASS / FAIL -------------------------------------------------
+    emit_verdict("tester", llm_verdict, detail=result, item_id=item.id)
+
+    if llm_verdict == "FAIL":
         item.test_fail_count += 1
         item.status = ItemStatus.IN_PROGRESS
         record_agent_outcome(state.repo_root, item.task_type, state.active_coder, success=False)
@@ -2214,20 +2311,126 @@ def tester_node(state: GraphState) -> dict:
     else:
         emit_status("tester", "✅ All tests PASSED", **_progress_meta(state, "deciding"))
 
-    emit_node_end("tester", "Testing", f"Verdict: {verdict}")
+    emit_node_end("tester", "Testing", f"Verdict: {llm_verdict}")
 
     updates = {
         "last_test_result": result,
-        "phase": WorkflowPhase.DECIDING if verdict == "PASS" else WorkflowPhase.CODING,
+        "env_fix_attempts": 0,  # reset on successful test run (pass or genuine fail)
+        "phase": WorkflowPhase.DECIDING if llm_verdict == "PASS" else WorkflowPhase.CODING,
     }
-    if verdict == "PASS":
+    if llm_verdict == "PASS":
         _save_checkpoint_snapshot(state, updates, "test_pass")
     return updates
 
 
 # ---------------------------------------------------------------------------
-# NODE: planner_decide
+# NODE: planner_env_fix
 # ---------------------------------------------------------------------------
+
+def planner_env_fix_node(state: GraphState) -> dict:
+    """Create a single env-setup fix item and prepend it to the plan.
+
+    Called when tester_node detects an environment failure (missing package,
+    command not found, etc.). This node:
+    1. Reads the tester's failure output from last_test_result
+    2. Uses an LLM to identify the exact missing package/tool
+    3. Creates one TodoItem: install the dependency or update config
+    4. Prepends it to todo_items so the coder handles it next
+    5. Routes to coder (skipping peer review — it's a mechanical fix)
+
+    Does NOT re-plan the entire task. Only creates the minimal fix item.
+    """
+    emit_node_start("planner", "EnvFix", item_desc="Diagnosing environment failure")
+    emit_status(
+        "planner",
+        "🔧 Planner: diagnosing missing dependency and creating fix item…",
+        **_progress_meta(state, "env_fixing"),
+    )
+
+    failure_output = state.last_test_result or "(no output captured)"
+    repo_facts_summary = ""
+    if state.repo_facts:
+        rf = state.repo_facts
+        lang = rf.get("tech_stack", {}).get("language", "unknown")
+        pm = rf.get("tech_stack", {}).get("package_manager", "unknown")
+        repo_facts_summary = f"Language: {lang}, Package manager: {pm}"
+
+    prompt = (
+        "## Environment Fix Task\n\n"
+        "The test runner failed due to a missing dependency or broken environment. "
+        "Your job is to create a single fix TodoItem to install or configure what is missing.\n\n"
+        f"**Test runner output (failure)**:\n```\n{failure_output[:3000]}\n```\n\n"
+        f"**Repository info**: {repo_facts_summary or 'unknown'}\n\n"
+        "Instructions:\n"
+        "1. Identify the exact missing package, module, or tool from the output above.\n"
+        "2. Determine the correct install command (pip install X, npm install X, "
+        "poetry add X, update pyproject.toml, etc.).\n"
+        "3. Respond with ONLY a JSON object in this exact format:\n"
+        "```json\n"
+        '{"description": "Install missing dependency X", '
+        '"command": "pip install X", '
+        '"reason": "ModuleNotFoundError: No module named X"}\n'
+        "```\n"
+        "Do not add explanation outside the JSON block."
+    )
+
+    raw = _invoke_agent("planner", [HumanMessage(content=prompt)], tools=None, inject_memory=False)
+
+    # Parse JSON from LLM response
+    fix_description = "Install missing test dependency"
+    fix_command = ""
+    try:
+        import json as _json
+        # Extract JSON block from response
+        json_start = raw.find("{")
+        json_end = raw.rfind("}") + 1
+        if json_start >= 0 and json_end > json_start:
+            parsed = _json.loads(raw[json_start:json_end])
+            fix_description = parsed.get("description", fix_description)
+            fix_command = parsed.get("command", "")
+    except Exception:
+        logger.warning("planner_env_fix | could not parse LLM JSON response: %s", raw[:200])
+
+    # Create the fix TodoItem with a unique id
+    fix_item = TodoItem(
+        id=f"env_fix_{state.env_fix_attempts + 1}",
+        description=fix_description,
+        task_type="ops",
+        acceptance_criteria=["Test environment is functional and tests can run"],
+        verification_commands=[fix_command] if fix_command else [],
+    )
+
+    # Assign the same coder that was active — env fixes skip peer review
+    fix_item.assigned_agent = state.active_coder
+    fix_item.assigned_reviewer = ""  # no review for mechanical env fixes
+
+    # Prepend fix item before the current item (so coder handles it next)
+    current_idx = state.current_item_index
+    new_items = (
+        list(state.todo_items[:current_idx])
+        + [fix_item]
+        + list(state.todo_items[current_idx:])
+    )
+
+    emit_status(
+        "planner",
+        f"🔧 Env fix item created: {fix_description}",
+        **_progress_meta(state, "env_fixing"),
+    )
+    emit_node_end("planner", "EnvFix", f"Fix item queued: {fix_description}")
+
+    return {
+        "todo_items": new_items,
+        "current_item_index": current_idx,  # points to fix_item now
+        "env_fix_attempts": state.env_fix_attempts + 1,
+        "phase": WorkflowPhase.CODING,
+        # Skip peer review for env-fix items — coder goes straight back to tester
+        "peer_review_verdict": "APPROVE",
+        "peer_review_notes": "Auto-approved: environment fix item, no peer review needed.",
+    }
+
+
+
 
 def planner_decide_node(state: GraphState) -> dict:
     """Mark item done, prepare commit."""

--- a/app/core/orchestrator.py
+++ b/app/core/orchestrator.py
@@ -20,6 +20,7 @@ from app.core.nodes import (
     learn_from_review_node,
     peer_review_node,
     planner_decide_node,
+    planner_env_fix_node,
     planner_plan_node,
     planner_review_node,
     research_node,
@@ -91,6 +92,8 @@ def _route_after_resume(state: GraphState) -> str:
         return "commit"
     if state.phase == WorkflowPhase.DOCUMENTING:
         return "documenter"
+    if state.phase == WorkflowPhase.ENV_FIXING:
+        return "env_fix"
     return "complete"
 
 
@@ -99,6 +102,11 @@ def _route_after_coder(state: GraphState) -> str:
         return "stopped"
     if state.phase == WorkflowPhase.WAITING_FOR_ANSWER:
         return "answer_gate"
+    # Env-fix items (ops tasks prepended by planner_env_fix) skip peer review
+    # and go straight back to tester so the fix can be verified immediately.
+    current = state.current_item
+    if current is not None and current.task_type == "ops" and current.id.startswith("env_fix_"):
+        return "tester"
     return "peer_review"
 
 
@@ -138,9 +146,22 @@ def _route_after_planner_review(state: GraphState) -> str:
 def _route_after_tester(state: GraphState) -> str:
     if _shutdown_event.is_set() or state.phase == WorkflowPhase.STOPPED:
         return "stopped"
+    if state.phase == WorkflowPhase.ENV_FIXING:
+        return "env_fix"
     if state.phase == WorkflowPhase.CODING:
         return "coder"
     return "decide"
+
+
+def _route_after_env_fix(state: GraphState) -> str:
+    """After planner_env_fix creates the fix item, always go to coder.
+
+    Env-fix items skip peer review — they are mechanical installs, not
+    creative coding tasks. The coder goes directly back to tester after.
+    """
+    if _shutdown_event.is_set() or state.phase == WorkflowPhase.STOPPED:
+        return "stopped"
+    return "coder"
 
 
 def _route_after_decide(state: GraphState) -> str:
@@ -193,6 +214,7 @@ def build_graph() -> StateGraph:
     graph.add_node("human_gate", human_gate_node)
     graph.add_node("commit", committer_node)
     graph.add_node("documenter", documenter_node)
+    graph.add_node("env_fix", planner_env_fix_node)
 
     graph.set_entry_point("router")
 
@@ -208,13 +230,13 @@ def build_graph() -> StateGraph:
     graph.add_conditional_edges(
         "resume",
         _route_after_resume,
-        {"coder": "coder", "answer_gate": "answer_gate", "commit": "commit", "documenter": "documenter", "complete": END, "stopped": END},
+        {"coder": "coder", "answer_gate": "answer_gate", "commit": "commit", "documenter": "documenter", "env_fix": "env_fix", "complete": END, "stopped": END},
     )
     graph.add_conditional_edges("planner", _route_after_plan, {"coder": "coder", "stopped": END})
     graph.add_conditional_edges(
         "coder",
         _route_after_coder,
-        {"answer_gate": "answer_gate", "peer_review": "peer_review", "stopped": END},
+        {"answer_gate": "answer_gate", "peer_review": "peer_review", "tester": "tester", "stopped": END},
     )
     graph.add_conditional_edges(
         "answer_gate",
@@ -232,7 +254,14 @@ def build_graph() -> StateGraph:
         _route_after_planner_review,
         {"tester": "tester", "coder": "coder", "stopped": END},
     )
-    graph.add_conditional_edges("tester", _route_after_tester, {"decide": "decide", "coder": "coder", "stopped": END})
+    graph.add_conditional_edges(
+        "tester",
+        _route_after_tester,
+        {"decide": "decide", "coder": "coder", "env_fix": "env_fix", "stopped": END},
+    )
+    graph.add_conditional_edges(
+        "env_fix", _route_after_env_fix, {"coder": "coder", "stopped": END}
+    )
     graph.add_conditional_edges("decide", _route_after_decide, {"human_gate": "human_gate", "stopped": END})
     graph.add_conditional_edges("human_gate", _route_after_human_gate, {"commit": "commit", "stopped": END})
     graph.add_conditional_edges("commit", _route_after_commit, {"documenter": "documenter", "complete": END})

--- a/app/core/state.py
+++ b/app/core/state.py
@@ -53,6 +53,7 @@ class WorkflowPhase(StrEnum):
     WAITING_FOR_ANSWER = "waiting_for_answer"
     COMMITTING = "committing"
     DOCUMENTING = "documenting"
+    ENV_FIXING = "env_fixing"   # planner is creating an env-setup fix item
     COMPLETE = "complete"
     STOPPED = "stopped"
 
@@ -149,6 +150,7 @@ class GraphState(BaseModel):
     # ── Metadata ──────────────────────────────────────────────────────
     total_iterations: int = 0
     completed_items: int = 0
+    env_fix_attempts: int = 0   # number of env-fix rounds attempted for current item
 
     @property
     def current_item(self) -> TodoItem | None:

--- a/tests/test_tester_env_failure.py
+++ b/tests/test_tester_env_failure.py
@@ -1,0 +1,388 @@
+"""Tests for tester env-failure detection and planner_env_fix routing (Issue #32).
+
+Covers:
+- _is_env_failure: all supported patterns
+- _is_test_pass: pass detection
+- _classify_test_output: full classification logic
+- tester_node: env_failure path → ENV_FIXING phase
+- tester_node: cap at _MAX_ENV_FIX_ATTEMPTS → STOPPED
+- tester_node: normal PASS path unchanged
+- tester_node: normal FAIL path unchanged
+- planner_env_fix_node: creates fix TodoItem, increments env_fix_attempts
+- planner_env_fix_node: prepends item to plan at correct index
+- Orchestrator routing: tester → env_fix, env_fix → coder
+- Orchestrator routing: coder → tester for env_fix items (skip peer review)
+- WorkflowPhase.ENV_FIXING present in enum
+- GraphState.env_fix_attempts default = 0
+"""
+from __future__ import annotations
+
+from unittest.mock import MagicMock, patch
+
+import pytest
+
+from app.core.state import GraphState, TodoItem, WorkflowPhase, ItemStatus
+
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+
+def _make_item(**kwargs) -> TodoItem:
+    defaults = dict(id="item_0", description="Add feature X")
+    defaults.update(kwargs)
+    return TodoItem(**defaults)
+
+
+def _make_state(**kwargs) -> GraphState:
+    defaults = dict(phase=WorkflowPhase.TESTING)
+    defaults.update(kwargs)
+    state = GraphState(**defaults)
+    return state
+
+
+# ---------------------------------------------------------------------------
+# 1. WorkflowPhase.ENV_FIXING and GraphState.env_fix_attempts
+# ---------------------------------------------------------------------------
+
+class TestStateAdditions:
+    def test_env_fixing_phase_exists(self):
+        assert hasattr(WorkflowPhase, "ENV_FIXING")
+        assert WorkflowPhase.ENV_FIXING == "env_fixing"
+
+    def test_env_fix_attempts_default_zero(self):
+        state = GraphState()
+        assert state.env_fix_attempts == 0
+
+    def test_env_fix_attempts_is_int(self):
+        state = GraphState(env_fix_attempts=2)
+        assert state.env_fix_attempts == 2
+
+
+# ---------------------------------------------------------------------------
+# 2. _is_env_failure pattern matching
+# ---------------------------------------------------------------------------
+
+class TestIsEnvFailure:
+    def _check(self, text: str) -> bool:
+        from app.core.nodes import _is_env_failure
+        return _is_env_failure(text)
+
+    def test_command_not_found(self):
+        assert self._check("bash: pytest: command not found") is True
+
+    def test_pytest_not_recognized_windows(self):
+        assert self._check("'pytest' is not recognized as an internal or external command") is True
+
+    def test_pytest_not_recognized_powershell(self):
+        assert self._check("The term 'pytest' is not recognized as the name of a cmdlet") is True
+
+    def test_no_module_named(self):
+        assert self._check("ModuleNotFoundError: No module named 'fastapi'") is True
+
+    def test_module_not_found_error(self):
+        assert self._check("ModuleNotFoundError: No module named 'uvicorn'") is True
+
+    def test_import_error(self):
+        assert self._check("ImportError: cannot import name 'Depends' from 'fastapi'") is True
+
+    def test_cannot_find_node_module(self):
+        assert self._check("Error: Cannot find module 'express'") is True
+
+    def test_externally_managed_environment(self):
+        assert self._check("error: externally-managed-environment") is True
+
+    def test_python_not_found(self):
+        assert self._check("python3: command not found") is True
+
+    def test_node_not_found(self):
+        assert self._check("node: not found") is True
+
+    def test_assertion_error_is_not_env_failure(self):
+        assert self._check("AssertionError: assert 1 == 2") is False
+
+    def test_test_failure_is_not_env_failure(self):
+        assert self._check("FAILED tests/test_api.py::test_health - assert response.status_code == 200") is False
+
+    def test_empty_string_is_not_env_failure(self):
+        assert self._check("") is False
+
+    def test_pass_output_is_not_env_failure(self):
+        assert self._check("5 passed in 1.23s") is False
+
+
+# ---------------------------------------------------------------------------
+# 3. _is_test_pass detection
+# ---------------------------------------------------------------------------
+
+class TestIsTestPass:
+    def _check(self, text: str) -> bool:
+        from app.core.nodes import _is_test_pass
+        return _is_test_pass(text)
+
+    def test_pytest_passed(self):
+        assert self._check("5 passed in 1.23s") is True
+
+    def test_all_tests_passed(self):
+        assert self._check("All tests passed") is True
+
+    def test_tests_passed(self):
+        assert self._check("Tests passed") is True
+
+    def test_fail_output_is_not_pass(self):
+        assert self._check("1 failed, 5 passed") is False
+
+    def test_empty_is_not_pass(self):
+        assert self._check("") is False
+
+
+# ---------------------------------------------------------------------------
+# 4. _classify_test_output
+# ---------------------------------------------------------------------------
+
+class TestClassifyTestOutput:
+    def _classify(self, text: str) -> str:
+        from app.core.nodes import _classify_test_output
+        return _classify_test_output(text)
+
+    def test_env_failure_wins_over_pass(self):
+        # Env failure takes priority
+        assert self._classify("ModuleNotFoundError: No module named 'x'\n5 passed") == "env_failure"
+
+    def test_pass_classified_correctly(self):
+        assert self._classify("5 passed in 0.5s") == "pass"
+
+    def test_test_failure_classified_correctly(self):
+        assert self._classify("1 failed, 3 passed\nAssertionError: assert 1 == 2") == "test_failure"
+
+    def test_empty_is_test_failure(self):
+        # Unknown output → test_failure (conservative)
+        assert self._classify("") == "test_failure"
+
+    def test_command_not_found_is_env_failure(self):
+        assert self._classify("pytest: command not found") == "env_failure"
+
+
+# ---------------------------------------------------------------------------
+# 5. tester_node env_failure path
+# ---------------------------------------------------------------------------
+
+class TestTesterNodeEnvFailure:
+    def _run(self, llm_result: str, env_fix_attempts: int = 0) -> dict:
+        from app.core.nodes import tester_node
+        from app.core.events import clear_listeners
+
+        item = _make_item()
+        state = _make_state(
+            todo_items=[item],
+            current_item_index=0,
+            env_fix_attempts=env_fix_attempts,
+            active_coder="coder_a",
+        )
+
+        with patch("app.core.nodes._invoke_agent", return_value=llm_result), \
+             patch("app.core.nodes._format_intelligence_summary_tester", return_value=""):
+            result = tester_node(state)
+        clear_listeners()
+        return result
+
+    def test_env_failure_sets_env_fixing_phase(self):
+        result = self._run("ModuleNotFoundError: No module named 'fastapi'")
+        assert result["phase"] == WorkflowPhase.ENV_FIXING
+
+    def test_env_failure_does_not_stop_workflow(self):
+        result = self._run("pytest: command not found")
+        assert result.get("stop_reason") != "env_setup_failed"
+
+    def test_env_failure_stores_last_test_result(self):
+        output = "ModuleNotFoundError: No module named 'uvicorn'"
+        result = self._run(output)
+        assert result["last_test_result"] == output
+
+    def test_env_failure_at_cap_stops_workflow(self):
+        from app.core.nodes import _MAX_ENV_FIX_ATTEMPTS
+        result = self._run(
+            "pytest: command not found",
+            env_fix_attempts=_MAX_ENV_FIX_ATTEMPTS,
+        )
+        assert result["phase"] == WorkflowPhase.STOPPED
+        assert result["stop_reason"] == "env_setup_failed"
+
+    def test_env_failure_below_cap_does_not_stop(self):
+        from app.core.nodes import _MAX_ENV_FIX_ATTEMPTS
+        result = self._run(
+            "pytest: command not found",
+            env_fix_attempts=_MAX_ENV_FIX_ATTEMPTS - 1,
+        )
+        assert result["phase"] == WorkflowPhase.ENV_FIXING
+
+    def test_pass_resets_env_fix_attempts(self):
+        result = self._run("5 passed in 0.5s\nVerdict: PASS")
+        assert result.get("env_fix_attempts") == 0
+
+    def test_test_failure_routes_to_coding(self):
+        result = self._run("1 failed\nVerdict: FAIL\nAssertionError")
+        assert result["phase"] == WorkflowPhase.CODING
+
+    def test_test_pass_routes_to_deciding(self):
+        result = self._run("5 passed\nVerdict: PASS")
+        assert result["phase"] == WorkflowPhase.DECIDING
+
+
+# ---------------------------------------------------------------------------
+# 6. planner_env_fix_node
+# ---------------------------------------------------------------------------
+
+class TestPlannerEnvFixNode:
+    def _run(self, last_test_result: str = "ModuleNotFoundError: No module named 'X'",
+             env_fix_attempts: int = 0,
+             items: list[TodoItem] | None = None,
+             current_index: int = 0) -> dict:
+        from app.core.nodes import planner_env_fix_node
+        from app.core.events import clear_listeners
+
+        if items is None:
+            items = [_make_item(id="item_0", description="Add feature")]
+
+        state = _make_state(
+            phase=WorkflowPhase.ENV_FIXING,
+            todo_items=items,
+            current_item_index=current_index,
+            last_test_result=last_test_result,
+            env_fix_attempts=env_fix_attempts,
+            active_coder="coder_a",
+            repo_facts={"tech_stack": {"language": "python", "package_manager": "pip"}},
+        )
+
+        llm_json = '{"description": "Install missing package X", "command": "pip install X", "reason": "ModuleNotFoundError"}'
+        with patch("app.core.nodes._invoke_agent", return_value=llm_json):
+            result = planner_env_fix_node(state)
+        clear_listeners()
+        return result
+
+    def test_returns_updated_todo_items(self):
+        result = self._run()
+        assert "todo_items" in result
+        assert len(result["todo_items"]) == 2  # original + fix item
+
+    def test_fix_item_prepended_before_current(self):
+        result = self._run()
+        fix_item = result["todo_items"][0]
+        assert fix_item.id.startswith("env_fix_")
+
+    def test_fix_item_task_type_ops(self):
+        result = self._run()
+        fix_item = result["todo_items"][0]
+        assert fix_item.task_type == "ops"
+
+    def test_current_item_index_unchanged(self):
+        result = self._run(current_index=0)
+        # index still points to position 0 which is now the fix item
+        assert result["current_item_index"] == 0
+
+    def test_env_fix_attempts_incremented(self):
+        result = self._run(env_fix_attempts=0)
+        assert result["env_fix_attempts"] == 1
+
+    def test_phase_set_to_coding(self):
+        result = self._run()
+        assert result["phase"] == WorkflowPhase.CODING
+
+    def test_peer_review_pre_approved(self):
+        # Env-fix items skip peer review — planner pre-approves
+        result = self._run()
+        assert result["peer_review_verdict"] == "APPROVE"
+
+    def test_fix_item_description_from_llm(self):
+        result = self._run()
+        fix_item = result["todo_items"][0]
+        assert "Install" in fix_item.description or len(fix_item.description) > 0
+
+    def test_multi_item_plan_fix_inserted_correctly(self):
+        items = [
+            _make_item(id="item_0", description="Feature A"),
+            _make_item(id="item_1", description="Feature B"),
+        ]
+        result = self._run(items=items, current_index=1)
+        # Fix item should be at index 1, Feature B at index 2
+        assert result["todo_items"][1].id.startswith("env_fix_")
+        assert result["todo_items"][2].id == "item_1"
+
+
+# ---------------------------------------------------------------------------
+# 7. Orchestrator routing
+# ---------------------------------------------------------------------------
+
+class TestEnvFixOrchestration:
+    @pytest.fixture(autouse=True)
+    def _reset_shutdown(self):
+        from app.core.orchestrator import reset_shutdown
+        reset_shutdown()
+        yield
+        reset_shutdown()
+
+    def test_tester_routes_to_env_fix_on_env_fixing_phase(self):
+        from app.core.orchestrator import _route_after_tester
+        state = _make_state(phase=WorkflowPhase.ENV_FIXING)
+        assert _route_after_tester(state) == "env_fix"
+
+    def test_tester_routes_to_decide_on_deciding_phase(self):
+        from app.core.orchestrator import _route_after_tester
+        state = _make_state(phase=WorkflowPhase.DECIDING)
+        assert _route_after_tester(state) == "decide"
+
+    def test_tester_routes_to_coder_on_coding_phase(self):
+        from app.core.orchestrator import _route_after_tester
+        state = _make_state(phase=WorkflowPhase.CODING)
+        assert _route_after_tester(state) == "coder"
+
+    def test_tester_routes_to_stopped_on_stopped_phase(self):
+        from app.core.orchestrator import _route_after_tester
+        state = _make_state(phase=WorkflowPhase.STOPPED)
+        assert _route_after_tester(state) == "stopped"
+
+    def test_env_fix_routes_to_coder(self):
+        from app.core.orchestrator import _route_after_env_fix
+        state = _make_state(phase=WorkflowPhase.CODING)
+        assert _route_after_env_fix(state) == "coder"
+
+    def test_env_fix_routes_to_stopped_on_shutdown(self):
+        from app.core.orchestrator import _route_after_env_fix, request_shutdown
+        request_shutdown()
+        state = _make_state(phase=WorkflowPhase.CODING)
+        assert _route_after_env_fix(state) == "stopped"
+
+    def test_coder_skips_peer_review_for_env_fix_item(self):
+        from app.core.orchestrator import _route_after_coder
+        fix_item = _make_item(id="env_fix_1", task_type="ops")
+        state = _make_state(
+            phase=WorkflowPhase.CODING,
+            todo_items=[fix_item],
+            current_item_index=0,
+        )
+        assert _route_after_coder(state) == "tester"
+
+    def test_coder_uses_peer_review_for_normal_items(self):
+        from app.core.orchestrator import _route_after_coder
+        normal_item = _make_item(id="item_0", task_type="coding")
+        state = _make_state(
+            phase=WorkflowPhase.CODING,
+            todo_items=[normal_item],
+            current_item_index=0,
+        )
+        assert _route_after_coder(state) == "peer_review"
+
+    def test_graph_contains_env_fix_node(self):
+        from app.core.orchestrator import build_graph
+        graph = build_graph()
+        assert "env_fix" in graph.nodes
+
+    def test_graph_compiles_with_env_fix(self):
+        from app.core.orchestrator import compile_graph
+        compiled = compile_graph()
+        assert compiled is not None
+
+    def test_resume_routes_to_env_fix_on_env_fixing_phase(self):
+        from app.core.orchestrator import _route_after_resume
+        state = _make_state(phase=WorkflowPhase.ENV_FIXING)
+        assert _route_after_resume(state) == "env_fix"

